src/Lean/Elab/Tactic/BVDecide/Frontend/Normalize/Simproc.lean

```diff
@@ -501,5 +501,64 @@ builtin_simproc [bv_normalize] extract_mul
       (← mkDecideProof (← mkLe lenExpr widthExpr))
   return .visit { expr := expr, proof? := some proof }
 
+/-!
+Lemmas about add-of-append are written using simp lemmas.
+This ensures that we correctly unify with `++`, even when the result type of the append
+has been reduced.
+If registered directly as a `simp` lemma, it would fail to unify against
+`(x : BitVec 3) ++ (y : BitVec 4)` with the result type `BitVec 7`.
+Simp would expect the resulting type to be `BitVec (3 + 4)`.
+-/
+builtin_simproc [bv_normalize] append_add_append_eq_append
+    ((HAppend.hAppend (α := BitVec (no_index _)) (β := BitVec (no_index _)) (γ := BitVec (no_index _)) _ _) +
+     (HAppend.hAppend (α := BitVec (no_index _)) (β := BitVec (no_index _)) (γ := BitVec (no_index _)) _ _))
+      := fun e => do
+  let_expr HAdd.hAdd _ _ _ _ lhsExpr rhsExpr := e | return .continue
+  let_expr HAppend.hAppend _ _ _ _ lhsExpr lhsZero := lhsExpr | return .continue
+  let_expr HAppend.hAppend _ _ _ _ rhsZero rhsExpr := rhsExpr | return .continue
+  let some ⟨w, lhsZeroValue⟩ ← getBitVecValue? lhsZero | return .continue
+  let some ⟨v, rhsZeroValue⟩ ← getBitVecValue? rhsZero | return .continue
+  if lhsZeroValue ≠ 0#w then
+    return .continue
+  if rhsZeroValue ≠ 0#v then
+    return .continue
+  let wExpr := toExpr w
+  let vExpr := toExpr v
+  let expr ← mkAppM ``HAppend.hAppend #[lhsExpr, rhsExpr]
+  let proof :=
+    mkApp4
+      (mkConst ``BitVec.append_add_append_eq_append)
+      vExpr
+      wExpr
+      lhsExpr
+      rhsExpr
+  return .visit { expr := expr, proof? := some proof }
+
+builtin_simproc [bv_normalize] append_add_append_eq_append'
+    ((HAppend.hAppend (α := BitVec (no_index _)) (β := BitVec (no_index _)) (γ := BitVec (no_index _)) _ _) +
+     (HAppend.hAppend (α := BitVec (no_index _)) (β := BitVec (no_index _)) (γ := BitVec (no_index _)) _ _))
+      := fun e => do
```

Comment:
I chose to perform code duplication here, because the two cases are different enough that I felt the code would be messier by unifying it into a single subroutine.